File: php/allergies/Allergies.php
<?php

declare(strict_types=1);

/**
 * Allergies
 *
 * This class represents a person's allergy profile based on an allergy score.
 * It uses bitwise operations to efficiently determine if a person is allergic to specific items
 * and to retrieve a list of all known allergies.
 */
class Allergies
{
    /**
     * @var int The numeric score representing the person's allergies.
     * Each bit in this integer corresponds to a specific allergen.
     */
    private int $score;

    /**
     * Constructs a new Allergies instance with a given allergy score.
     *
     * @param int $score The integer representing the person's allergy profile.
     */
    public function __construct(int $score)
    {
        $this->score = $score;
    }

    /**
     * Checks if the person is allergic to a specific allergen.
     *
     * This is determined by performing a bitwise AND operation between the allergen's score
     * and the person's total allergy score. If the result is non-zero, it indicates an allergy.
     *
     * @param Allergen $allergen An instance of Allergen representing the allergen to check.
     * @return bool True if the person is allergic to the given allergen, false otherwise.
     *
     * @example
     * $allergies = new Allergies(3); // Score 3 means allergic to Eggs (1) and Peanuts (2)
     * $allergies->isAllergicTo(new Allergen(Allergen::EGGS));    // Returns true
     * $allergies->isAllergicTo(new Allergen(Allergen::PEANUTS)); // Returns true
     * $allergies->isAllergicTo(new Allergen(Allergen::CATS));    // Returns false
     */
    public function isAllergicTo(Allergen $allergen): bool
    {
        return ($allergen->getScore() & $this->score) !== 0;
    }

    /**
     * Retrieves a list of all allergens the person is allergic to.
     *
     * It iterates through a static list of all possible allergens and filters
     * them based on the `isAllergicTo` method.
     *
     * @return array<Allergen> An array of Allergen objects to which the person is allergic.
     *
     * @example
     * $allergies = new Allergies(3); // Allergic to Eggs and Peanuts
     * // $list will contain Allergen objects for Eggs and Peanuts.
     * // Example check: array_map(fn($a) => $a->getScore(), $allergies->getList()) could return [1, 2]
     *
     * $allergies = new Allergies(0); // Allergic to nothing
     * $allergies->getList(); // Returns []
     */
    public function getList(): array
    {
        return array_filter(Allergen::allergenList(), fn($allergen) => $this->isAllergicTo($allergen));
    }

    /**
     * Returns the raw allergy score.
     *
     * @return int The integer score representing the person's allergy profile.
     */
    public function getScore(): int
    {
        return $this->score;
    }
}

/**
 * Allergen
 *
 * This class represents a single allergen with its corresponding score value.
 * Each allergen is assigned a unique power-of-2 score, allowing for bitwise
 * combination in an allergy profile.
 */
class Allergen
{
    /**
     * @var int EGGS Score value for Eggs.
     */
    const EGGS = 1; // 2^0
    /**
     * @var int PEANUTS Score value for Peanuts.
     */
    const PEANUTS = 2; // 2^1
    /**
     * @var int SHELLFISH Score value for Shellfish.
     */
    const SHELLFISH = 4; // 2^2
    /**
     * @var int STRAWBERRIES Score value for Strawberries.
     */
    const STRAWBERRIES = 8; // 2^3
    /**
     * @var int TOMATOES Score value for Tomatoes.
     */
    const TOMATOES = 16; // 2^4
    /**
     * @var int CHOCOLATE Score value for Chocolate.
     */
    const CHOCOLATE = 32; // 2^5
    /**
     * @var int POLLEN Score value for Pollen.
     */
    const POLLEN = 64; // 2^6
    /**
     * @var int CATS Score value for Cats.
     */
    const CATS = 128; // 2^7

    /**
     * Generates a static list of all possible Allergen objects.
     *
     * Each allergen in the list is created with its corresponding power-of-2 score.
     *
     * @return array<Allergen> An array containing an instance for each defined allergen.
     */
    public static function allergenList(): array
    {
        return array_map(fn(int $exp) => new Allergen(2 ** $exp), range(0, 7));
    }

    /**
     * @var int The specific score value of this allergen.
     */
    private int $score;

    /**
     * Constructs a new Allergen instance with a specific score.
     *
     * @param int $score The integer score of the allergen (e.g., 1 for Eggs, 2 for Peanuts).
     */
    public function __construct(int $score)
    {
        $this->score = $score;
    }

    /**
     * Returns the score value of this allergen.
     *
     * @return int The score of the allergen.
     */
    public function getScore(): int
    {
        return $this->score;
    }
}

File: lua/allergies/allergies.lua
--- A module for checking and listing allergies based on a numerical score (bitmask).
local allergies <const> = {}

--- An enumeration of possible allergies, with each allergy assigned a unique bit value.
--- Each value is a power of two, allowing multiple allergies to be stored in a single integer using bitwise operations.
--- @enum Allergy
local Allergy <const> = {
    eggs = { "eggs", 1 },
    peanuts = { "peanuts", 2 },
    shellfish = { "shellfish", 4 },
    strawberries = { "strawberries", 8 },
    tomatoes = { "tomatoes", 16 },
    chocolate = { "chocolate", 32 },
    pollen = { "pollen", 64 },
    cats = { "cats", 128 }
}

--- A constant table containing the names of all defined allergies in a specific order.
--- This list is used to ensure the `list` function returns results in a consistent, predictable order.
local ordereds_allergies <const> = {
    Allergy.eggs[1], Allergy.peanuts[1], Allergy.shellfish[1],
    Allergy.strawberries[1], Allergy.tomatoes[1], Allergy.chocolate[1],
    Allergy.pollen[1], Allergy.cats[1]
}

--- Returns a list of all allergies corresponding to a given score.
--- The function iterates through all known allergies and checks if each one is present in the score.
--- @param score integer The bitmask representing the user's allergies.
--- @return string[] -- A table of strings, where each string is an allergy name.
function allergies.list(score)
    local result = {}

    for _, item in ipairs(ordereds_allergies) do
        if allergies.allergic_to(score, item) then
            table.insert(result, item)
        end
    end

    return result
end

--- Checks if a specific allergy is present in the given score.
--- This check is performed using a bitwise AND operation. A non-zero result indicates that the bit for the specific allergy is set in the score.
--- @param score integer The bitmask representing the user's allergies.
--- @param which string The name of the allergy to check.
--- @return boolean -- True if the allergy is present in the score, false otherwise.
function allergies.allergic_to(score, which)
    return score & Allergy[which][2] ~= 0
end

return allergies